Introduce `profiles` for services

acran opened the pull request:
**What this PR does / why we need it**:
Introduce `profiles` in compose schema and specification.
This is based on the PR #98 by @ndeloof and the discussion in #97. It diverges from #98 by:
- limit initial scope to services only
- adds a regex for valid profile names (see https://github.com/compose-spec/compose-spec/pull/98#issuecomment-685732572)
- clarify that the profiles of services explicitly targeted by commands must be automatically enabled (see https://github.com/compose-spec/compose-spec/issues/97#issuecomment-685712571 and https://github.com/compose-spec/compose-spec/issues/97#issuecomment-683398226)
- added another example to demonstrate the last point

Automatically enabling the profiles of explicitly targeted service is the minimal amount of "magic" required here to allow for a sane UX when targeting services with dependencies in non-default profiles. But at the same time it still forces the users to specify a valid compose model where all dependencies share a common profile.

**Which issue(s) this PR fixes**:
Fixes #97
see also #98
Fixes #87

ndeloof commented on spec.md:
```diff
@@ -167,6 +167,57 @@ merged are hosted in other folders.
+implementation SHOULD allow the user to define a set of active profiles. The exact mechanism is implementation
+specific and MAY include command line flags, environment variables, etc.
+
+The Services top-level element supports a `profiles` attribute to define a list of named profiles. Services without
```
imho this should not be restricted to `services`, but applied to all top-level elements

acran replied:
For _this_ PR I reduced it to services to keep the scope as small as possible while it also already covers a majority of the use cases with the intention to expand with future PRs :+1:

EricHripko replied:
I think that's a great idea to limit the scope 🙌 To avoid ambiguity, should we clarify what happens to other objects in this case? I guess this can be as simple as specifying that all networks/volumes/etc are treated as always enabled for now.

EricHripko replied:
`All other top-level elements are not affected by `profiles` and are always active.`

This is now addressed 👍

EricHripko commented on spec.md:
```diff
@@ -167,6 +167,57 @@ merged are hosted in other folders.
+explicitly targeted by a command. In that case its `profiles` MUST be added to the set of active profiles.
+
+References to other services (by `links`, `extends` or shared resource syntax `service:xxx`) MUST not
+automatically enable a component that would otherwise have been ignored by active profiles.
```
perhaps worth adding that it should also produce an error?

EricHripko commented on spec.md:
````diff
@@ -167,6 +167,57 @@ merged are hosted in other folders.
+```
+
+- Compose application model parsed with no profile enabled only contains the `foo` service.
+- If profile `test` is enabled, model contains the services `foo`, `bar` and `baz`.
````
I think this is a great example 🎉  Let's explain why each service is pulled in for clarity's sake:
- `foo` = no profiles = always enabled
- `bar` + `baz` = one of profiles matches active profile

EricHripko commented on spec.md:
```diff
@@ -167,6 +167,57 @@ merged are hosted in other folders.
+
+- Compose application model parsed with no profile enabled only contains the `foo` service.
+- If profile `test` is enabled, model contains the services `foo`, `bar` and `baz`.
+- If profile `debug` is enabled, model contains both `foo` and `zot` services, but not `bar` and `baz`
```
Same as above

EricHripko commented on spec.md:
```diff
@@ -167,6 +167,57 @@ merged are hosted in other folders.
+- If profile `test` is enabled, model contains the services `foo`, `bar` and `baz`.
+- If profile `debug` is enabled, model contains both `foo` and `zot` services, but not `bar` and `baz`
+  and as such the model is invalid regarding the `depends_on` constraint of `zot`.
+- If Compose implementation is executed with `bar` as explicit service to run, it will be active
```
I think it's worth specifying that `test` profile will be enabled implicitly as well.
It illustrates the spec a bit better and means that, for example, `compose up bar` will also start `baz` (`tests` in the profile) and `foo` (always enabled).

acran replied:
actually no, `compose up bar` will only start `bar` since it has no dependencies to `foo` or `baz` (both would be in the active profiles though, just not started by the command)

EricHripko replied:
This sounds inconsistent. On one hand, the PR says:

> All other top-level elements are not affected by `profiles` and are always active.

On the other hand, you suggest that:

> both would be in the active profiles though, just not started by the command

acran replied:
Sorry, I can not quite follow here:
_All other top-level elements_ is intended to mean _non-service top-level elements_ like _networks_, _volumes_ etc. but `foo` and `baz` are services, so this does not apply here?

With `compose up bar` the services `foo` and `baz` will be _included_ in the compose model for the reasons you mentioned, but they will not be _started_ since only the service `bar` is targeted by the command explicitly^^

EricHripko replied:
Ah sorry, I confused myself here 🙈 Thank you for clarifying!

EricHripko commented on spec.md:
```diff
@@ -167,6 +167,57 @@ merged are hosted in other folders.
+  and as such the model is invalid regarding the `depends_on` constraint of `zot`.
+- If Compose implementation is executed with `bar` as explicit service to run, it will be active
+  even if `test` profile is not enabled by the user.
+- If Compose implementation is executed with `baz` as explicit service to run, the service `baz` will
```
Same as above

EricHripko commented on spec.md:
```diff
@@ -167,6 +167,57 @@ merged are hosted in other folders.
+- If Compose implementation is executed with `zot` as explicit service to run, again the model will be
+  invalid regarding the `depends_on` constraint of `zot` since `zot` and `bar` have no common `profiles`
+  listed.
+
```
Given that spec suggests that multiple profiles can be active, I think it's worth adding an example for that

EricHripko reviewed (commented):
Looks like an awesome start 🙌 I've left a few comments inline 💬

acran commented:
I updated the PR with @EricHripko comments and added/clarified some examples.
But it already feels a bit bloated with that many examples and like over-explaining every single case in detail?

acran commented:
So, any further comments on the updated PR, or can this be moved forward?

EricHripko commented on spec.md:
```diff
@@ -167,6 +167,64 @@ merged are hosted in other folders.
+specific and MAY include command line flags, environment variables, etc.
+
+The Services top-level element supports a `profiles` attribute to define a list of named profiles. Services without
+a `profiles` attribute set MUST always be enabled. A service MUST be ignored by the compose
```
```suggestion
a `profiles` attribute set MUST always be enabled. A service MUST be ignored by the Compose
```
nit: In the spec `Compose` starts with a capital letter

EricHripko commented on spec.md:
```diff
@@ -167,6 +167,64 @@ merged are hosted in other folders.
+- If Compose implementation is executed with `zot` as explicit service to run, again the model will be
+  invalid regarding the `depends_on` constraint of `zot` since `zot` and `bar` have no common `profiles`
+  listed.
+- If Compose implementation is executed with `zot` as explicit service to run and peofile `test` enabled,
```
```suggestion
- If Compose implementation is executed with `zot` as explicit service to run and profile `test` enabled,
```
Typo

EricHripko reviewed (commented):
2 minor wording suggestions - otherwise this looks good to go 🙌

acran commented:
Thanks @EricHripko 
I now fixed the two typos and also proof-read again, should be fine now :tada:

EricHripko reviewed (approved):
Looks great ✅ Thank you for picking up and finishing the initial draft of this proposal 👍
